### PR TITLE
hardening/3_add_cache_for_auth_tokens

### PR DESCRIPTION
**Pull request by pcoello25**
- Fix issue #3 
- e2e (unofficial) tests passed:

```
[ VALID USER & TOKEN -> NOT CACHED]
info: Starting cosmos-proxy in <host>:<port>
info: Authentication OK: {"organizations": [], "displayName": "pcoello25", "roles": [], "app_id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", "isGravatarEnabled": false, "email": "<email_adress>", "id": "pcoello25"}
info: Authorization OK: user pcoello25 is allowed to access /webhdfs/v1/user/pcoello25
info: Redirecting to http://<host>:<port>

[ VALID USER & TOKEN -> USER IN CACHE -> TOKEN NOT CACHED]
info: Starting cosmos-proxy in <host>:<port>
info: Authentication OK: {"organizations": [], "displayName": "pcoello25", "roles": [], "app_id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", "isGravatarEnabled": false, "email": "<email_adress>", "id": "pcoello25"}
info: Authorization OK: user pcoello25 is allowed to access /webhdfs/v1/user/pcoello25
info: Redirecting to http://<host>:<port>

[ VALID USER & TOKEN -> USER IN CACHE -> TOKEN IN CACHE]
info: Starting cosmos-proxy in <host>:<port>
info: Authorization OK: user pcoello25 is allowed to access /webhdfs/v1/user/pcoello25
info: Redirecting to http://<host>:<port>

[VALID USER & INVALID TOKEN]
info: Authentication OK: {"organizations": [], "displayName": "pcoello25", "roles": [], "app_id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", "isGravatarEnabled": false, "email": "<email_adress>", "id": "pcoello25"}
error: Authorization error: user pcoello25 is not allowed to access /webhdfs/v1/user/frb

[UNREGISTERED USER]
error: Authentication error: {"error": {"message": "Access Token faketoken not found", "code": 404, "title": "Not Found"}}

[ACCESS TO PUBLIC PATHS WITH A VALID USER AND A INVALID USER]
info: Starting cosmos-proxy in <host>:<port>
info: Authentication OK: {"organizations": [], "displayName": "pcoello25", "roles": [], "app_id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", "isGravatarEnabled": false, "email": "<email_adress>", "id": "pcoello25"}
error: Authorization error: user pcoello25 is not allowed to access /webhdfs/v1/user/frb/opendata
info: Authentication OK: {"organizations": [], "displayName": "pcoello25", "roles": [], "app_id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", "isGravatarEnabled": false, "email": "<email_adress>", "id": "pcoello25"}
info: Authorization OK: user pcoello25 is allowed to access /webhdfs/v1/user/pcoello25/opendata
info: Redirecting to http://<host>:<port>
```
- Assignee: @frbattid

### CODE REVIEW

**Comment by pcoello25** [comment]
Added new funcionality: `cache-dump.json` file support in case of proxy shutdown. There are a log added for tests purpose similar to ' user is trying to access to path with token token... '. It will be removed with the last changes in the code.
- New e2e (unofficial) tests:

```
[VALID USER & TOKEN --> EMPTY CACHE]
curl -X GET "http://localhost:14000/webhdfs/v1/user/pcoello25?op=liststatus&user.name=pcoello25" -H "X-Auth-Token: pc25token"

info: Starting cosmos-proxy in 0.0.0.0:14000
info: Empty 'cache-dump' file. Starting an empty cache.
info: pcoello25 is trying to access to /webhdfs/v1/user/pcoello25 with the token pcoello25token
info: Authentication OK: {"organizations": [], "displayName": "pcoello25", "roles": [], "app_id": "4d1af2eec3754099a4f8dc86bf735068", "isGravatarEnabled": false, "email": "pablo.coellovillalba@telefonica.com", "id": "pcoello25"}
info: Authorization OK: user pcoello25 is allowed to access /webhdfs/v1/user/pcoello25


[VALID USER & TOKEN --> USER & TOKEN IN CACHE]
curl -X GET "http://localhost:14000/webhdfs/v1/user/pcoello25?op=liststatus&user.name=pcoello25" -H "X-Auth-Token: pcoello25token"

info: Starting cosmos-proxy in 0.0.0.0:14000
info: Loading cache from cache-dump file
info: pcoello25 is trying to access to /webhdfs/v1/user/pcoello25 with the token pcoello25token
info: Authorization OK: user pcoello25 is allowed to access /webhdfs/v1/user/pcoello25


[UNKNOWN USER & TOKEN]
curl -X GET "http://localhost:14000/webhdfs/v1/user/unknownuser?op=liststatus&user.name=unknownuser" -H "X-Auth-Token: unknowntoken"

info: Starting cosmos-proxy in 0.0.0.0:14000
info: 'cache-dump' file not found. Starting an empty cache.
info: unknownUser is trying to access to /webhdfs/v1/user/unknownUser with the token unknownToken
error: Authentication error: {"error": {"message": "Access Token unknownToken not found", "code": 404, "title": "Not Found"}}


[VALID USER & INVALID TOKEN --> BOTH CASES: EMTPY CACHE & CACHE WITH INFO]
curl -X GET "http://localhost:14000/webhdfs/v1/user/pcoello25?op=liststatus&user.name=pcoello25" -H "X-Auth-Token: unknownToken"

info: Starting cosmos-proxy in 0.0.0.0:14000
info: Empty 'cache-dump' file. Starting an empty cache.
info: pcoello25 is trying to access to /webhdfs/v1/user/pcoello25 with the token unknownToken
error: Authentication error: {"error": {"message": "Access Token unknownToken not found", "code": 404, "title": "Not Found"}}


[INVALID USER & VALID TOKEN--> BOTH CASES: EMTPY CACHE & CACHE WITH INFO]
curl -X GET "http://localhost:14000/webhdfs/v1/user/frb?op=liststatus&user.name=frb" -H "X-Auth-Token: pcoello25token"

info: Starting cosmos-proxy in 0.0.0.0:14000
info: File doesn't exist
info: cache-dump file not found. Starting an empty cache.
info: frb is trying to access to /webhdfs/v1/user/frb with the token ZySUCQYO8eZjvrtsBG4XCOg9jqIT1s
error: Authentication error: {"error": {"message": "User doesn't match the provided token","code": 404, "title": "Not authenticated"}}


[VALID USER & TOKEN --> INVALID PATH -> EMPTY CACHE]
curl -X GET "http://localhost:14000/webhdfs/v1/user/frb?op=liststatus&user.name=pcoello25" -H "X-Auth-Token: pcoello25token"

info: Starting cosmos-proxy in 0.0.0.0:14000
info: 'cache-dump' file not found. Starting an empty cache.
info: pcoello25is trying to access to /webhdfs/v1/user/frb with the token pcoello25token
info: Authentication OK: {"organizations": [], "displayName": "pcoello25", "roles": [], "app_id": "4d1af2eec3754099a4f8dc86bf735068", "isGravatarEnabled": false, "email": "pablo.coellovillalba@telefonica.com", "id": "pcoello25"}
error: Authorization error: user pcoello25 is not allowed to access /webhdfs/v1/user/frb


[VALID USER & TOKEN --> INVALID PATH --> USER & TOKEN EN CACHE]
curl -X GET "http://localhost:14000/webhdfs/v1/user/frb?op=liststatus&user.name=pcoello25" -H "X-Auth-Token: CKQ4IFamATtPq2T3Iz5kUP9AB2h2pe"

info: Starting cosmos-proxy in 0.0.0.0:14000
info: Loading cache from cache-dump file
info: pcoello25is trying to access to /webhdfs/v1/user/frb with the token CKQ4IFamATtPq2T3Iz5kUP9AB2h2pe
error: Authorization error: user pcoello25 is not allowed to access /webhdfs/v1/user/frb
```

**Comment by pcoello25** [comment]
In the last commit ( 2869574) it has been fixed the log when a user is trying to access to his path using the token of another user. In addition, it has been added 2 new classes for improve the code.

**Comment by frbattid** [comment]
CHANGES_NEXT_RELEASE update is required.

**Comment by frbattid** [comment]
Explanation about the cache file is required in the README.md ;)

File: src/cache.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ */
+
+var fs = require('fs'),
+    cache = [],
+    conf = require('../conf/cosmos-proxy.json');
+
+var pathToFile = conf.cache_file;
+
+function createEmptyFileCache() {
+    return fs.closeSync(fs.openSync(pathToFile,'w'));
```
Comment (frbattid) [inline src/cache.js]: White space after comma `,` is required.

File: src/cache.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+} // createEmptyCache
+
+function isCacheEmpty() {
+    return (cache == null || cache.length == undefined || cache.length == 0);
+} // isCacheEmpty
+
+function isInCache(reqUser) {
+    if (!isCacheEmpty()) {
+        for (var i=0; i<cache.length; i++) {
+            for (var key in cache[i]) {
```
Comment (frbattid) [inline src/cache.js]: This part:

```
for (var key in cache[i]) {
                if (cache[i].hasOwnProperty(key)) {
                    if ((key === 'user') && (cache[i][key] === reqUser)) {
                        return true;
                    } // if
                } // if
```

Should be changed by something like:

```
var exists = false;

for (var i = 0; i < cache.length; i++) {
   if (cache[i]['user'] != reqUser) {
       exists = true;
       break;
   }
}

return exists;
```

File: src/cache.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            } // for
+        } // for
+    } // if
+    return false;
+} // isInCache
+
+function isUsersToken(reqUser, token) {
+    var isCachedUser = false;
+    if (!isCacheEmpty()) {
+        for (var i=0; i<cache.length; i++) {
```
Comment (frbattid) [inline src/cache.js]: This part should be:

```
var matches = false;

for (var i = 0; i < cache.length; i++) {
   if (cache[i]['user'] != reqUser && cache[i]['token`] !== token) {
       matches = true;
       break;
   }
}

return matches;
```

File: src/cache.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                } // if
+            } // for
+            isCachedUser = false;
+        } // for
+    } // if
+    return false;
+} // isInCache
+
+function isCacheAuthenticated(reqUser, token) {
+    if (isInCache(reqUser)) {
```
Comment (frbattid) [inline src/cache.js]: `isInCache` is not necessary here.

File: src/cache.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+function createEmptyFileCache() {
+    return fs.closeSync(fs.openSync(pathToFile,'w'));
+} // createEmptyCache
+
+function isCacheEmpty() {
+    return (cache == null || cache.length == undefined || cache.length == 0);
+} // isCacheEmpty
+
+function isInCache(reqUser) {
```
Comment (frbattid) [inline src/cache.js]: Finally, we have found this function is not necessary.

File: src/tidoopfs.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        return true;
+    } catch (e) {
+        return false;
+    } // try catch
+} // fileExists
+
+module.exports = {
+    isEmptyFile: isEmptyFile,
+    fileExists: fileExists
+} // module.exports
```
Comment (frbattid) [inline src/tidoopfs.js]: Final white line is required.

File: src/cosmos-proxy.js [hunk 7]
```diff
@@ -27,10 +27,30 @@ var http = require('http'),
     idm = require('./idm.js'),
     logger = require('./logger.js'),
     conf = require('../conf/cosmos-proxy.json');
+    cache = require('./cache.js');
+    tidoopfs = require('./tidoopfs.js');
 
 logger.info('Starting cosmos-proxy in ' + conf.host + ':' + conf.port);
 var proxy = httpProxy.createProxyServer({});
 
+if (cache.isCacheEmpty()) {
```
Comment (frbattid) [inline src/cosmos-proxy.js:37]: Just add a comment saying something like "Unnecessary, but do not disturb"

File: src/cosmos-proxy.js [hunk 8]
```diff
@@ -41,47 +61,77 @@ function isWhiteListed(list, path) {
     return false;
 } // isWhiteListed
 
+function isAuthorized(username, path) {
+    var whiteListed = isWhiteListed(conf.public_paths_list, path);
```
Comment (frbattid) [inline src/cosmos-proxy.js:66]: Whte line before the `if` is required.

File: src/cosmos-proxy.js [hunk 9]
```diff
@@ -41,47 +61,77 @@ function isWhiteListed(list, path) {
...
     var token = req.headers['x-auth-token'];
 
-    idm.authenticate(token, function(error, result) {
-        if (error) {
-            logger.error('Authentication error: ' + error);
-            res.writeHead(400, { 'Content-Type': 'text/plain' });
-            res.end('Authentication error: ' + error);
-        } else {
-            var json = JSON.parse(result);
+    logger.info(reqUser + ' is trying to access to ' + path + ' with the token ' + token + '......');
```
Comment (frbattid) [inline src/cosmos-proxy.js]: This is an internal log, you can trace the token.

File: src/cosmos-proxy.js [hunk 10]
```diff
@@ -41,47 +61,77 @@ function isWhiteListed(list, path) {
...
                 } else {
-                    if (user === conf.superuser) {
-                        logger.info('Authorization OK: user ' + user + ' is allowed to access ' + path);
-                        logger.info('Redirecting to http://' + conf.target.host + ':' + conf.target.port);
-                        proxy.web(req, res, {target: 'http://' + conf.target.host + ':' + conf.target.port}); // forward to the target server
-                    } else if (path.indexOf('/webhdfs/v1/user/' + user) == 0) {
-                        logger.info('Authorization OK: user ' + user + ' is allowed to access ' + path);
-                        logger.info('Redirecting to http://' + conf.target.host + ':' + conf.target.port);
-                        proxy.web(req, res, {target: 'http://' + conf.target.host + ':' + conf.target.port}); // forward to the target server
+                    var idmUser = json['id'];
```
Comment (frbattid) [inline src/cosmos-proxy.js:113]: White line is required after this one.

Comment (pcoello25) [reply src/cosmos-proxy.js:113]: Fixed in 5c0f100

File: src/cosmos-proxy.js (hunk 9 above)
Comment (pcoello25) [reply src/cosmos-proxy.js]: Fixed in 5c0f100

File: src/cosmos-proxy.js (hunk 8 above)
Comment (pcoello25) [reply src/cosmos-proxy.js:66]: Fixed in 5c0f100

File: src/cosmos-proxy.js (hunk 7 above)
Comment (pcoello25) [reply src/cosmos-proxy.js:37]: Fixed in 5c0f100

File: src/tidoopfs.js (hunk 6 above)
Comment (pcoello25) [reply src/tidoopfs.js]: Fixed in 5c0f100

File: src/cache.js (hunk 5 above)
Comment (pcoello25) [reply src/cache.js]: Fixed in 5c0f100

File: src/cache.js (hunk 4 above)
Comment (pcoello25) [reply src/cache.js]: Fixed in 5c0f100

File: src/cache.js (hunk 3 above)
Comment (pcoello25) [reply src/cache.js]: Fixed in 5c0f100

File: src/cache.js (hunk 2 above)
Comment (pcoello25) [reply src/cache.js]: Fixed in 5c0f100

File: src/cache.js (hunk 1 above)
Comment (pcoello25) [reply src/cache.js]: Fixed in 5c0f100

**Comment by frbattid** [comment]
LGTM

**Pull request merged**